### PR TITLE
Add enif_dynamic_resource_call

### PR DESCRIPTION
**Pull request by sverker**
As a "safe" way to call NIF code in another module.

Also add `enif_init_resource_type` as a new (and hopefully the final) way to create a resource type with the new callback function `dyncall`.

### CODE REVIEW

File: erts/emulator/test/nif_SUITE_data/nif_mod.c [hunk 1]
```diff
@@ -161,15 +175,31 @@ static void open_resource_type(ErlNifEnv* env, int arity, const ERL_NIF_TERM* ar
...
-            init.down = resource_down_D;
+	init.down = NULL;
+
+#  ifdef HAVE_ENIF_DYNAMIC_RESOURCE_CALL
+	init.members = 0xdead;
+	init.dyncall = (ErlNifResourceDynCall*) 0xdeadbeaf;
+
+	if (enif_is_identical(arr[6], am_resource_dyncall)) {
+            init.dyncall = resource_dyncall;
+	    init.members = 4;
```
Comment (garazdawi) [inline erts/emulator/test/nif_SUITE_data/nif_mod.c:186]: Should we provide a macro that does this initialization? i.e. something 

```c
#define ERL_NIF_INIT_RESOURCE_TYPE(init) \
  memset(init,0,sizeof(ErlNifResourceTypeInit)); \
  init->members = CURRENT_MAX_MEMBERS
```

Comment (sverker) [reply erts/emulator/test/nif_SUITE_data/nif_mod.c:186]: Yes maybe, but that doesn't help if you have a static initialized struct

`ErlNifResourceTypeInit my_resource_init = { my_dtor, ... };
`

Comment (garazdawi) [reply erts/emulator/test/nif_SUITE_data/nif_mod.c:186]: true...

**Comment by josevalim** [comment]
Is the goal here to implement some sort of callback? As a pseudo example, the inet module could say it works with different types of NIF resources and it expects them to implement certain dyncall?

**Comment by RaimoNiskanen** [comment]
@josevalim: The origin of this feature is the need to get a file descriptor from the file NIF to be used in sendfile from the socket NIF, in an atomic way so to not leak it nor give it a chance to change target during the operation.

I foresee that some other NIF also might want to get a file descriptor from the file NIF, so I sketched on an API there, thinking about forward/backward compatibility.  I should/will create a PR of the ongoing sendfile work, now that I missed 24.0-rc3, so we can discuss how dynamic calls should be used...

**Comment by lhoguin** [comment]
Back when I started working on `esdl2` (SDL2 NIF), I wanted to have a separate NIF for each of the SDL2 and related components (SDL2 itself, SDL_Image, SDL_ttf, OpenGL and others). This way someone that only wants SDL2 could avoid including OpenGL for example. _Bonus points if I could use the OpenGL available with `wx` instead of writing it, as well._

This required passing a resource to the other NIF and being able to extract information from it (some kind of handle, window, image and such). But it did not seem possible at the time. It sounds like this PR would allow doing exactly that.

**Review by RaimoNiskanen — COMMENTED** [review]
Gr8!

**Pull request merged**